pkg/s3signer/request-signature-streaming.go

```diff
@@ -0,0 +1,260 @@
+	if dataLen <= 0 {
+		return 0
+	}
+	chunksCount := int64(dataLen / chunkSize)
+	remainingBytes := int64(dataLen % chunkSize)
+	streamLen := int64(0)
+	streamLen += chunksCount * getSignedChunkLength(chunkSize)
+	if remainingBytes > 0 {
+		streamLen += getSignedChunkLength(remainingBytes)
+	}
+	streamLen += getSignedChunkLength(0)
+	return streamLen
+}
+
+// getChunkStringToSign - returns the string to sign given chunk data and previous signature.
+func getChunkStringToSign(t time.Time, location, previousSig string, chunkData []byte) string {
+	stringToSignParts := []string{
+		streamingPayloadHdr,
+		t.Format(iso8601DateFormat),
+		getScope(location, t),
+		previousSig,
+		emptySHA256,
+		hex.EncodeToString(sum256(chunkData)),
+	}
+
+	return strings.Join(stringToSignParts, "\n")
+}
+
+// PrepareStreamingRequest - prepares a request with appropriate headers before computing the seed signature.
+func PrepareStreamingRequest(req http.Request, dataLen int64, timestamp time.Time) (*http.Request, int64, error) {
+	// Set x-amz-content-sha256 header.
+	req.Header.Set("X-Amz-Content-Sha256", streamingSignAlgorithm)
+	req.Header.Set("Content-Encoding", streamingEncoding)
+	req.Header.Set("X-Amz-Date", timestamp.Format(iso8601DateFormat))
+
+	streamContentLen := getStreamLength(dataLen, int64(payloadChunkSize))
+	// Set content length with streaming signature for each chunk included.
+	req.Header.Set("Content-Length", strconv.FormatInt(streamContentLen, 10))
+	req.Header.Set("x-amz-decoded-content-length", strconv.FormatInt(dataLen, 10))
+
+	return &req, dataLen, nil
+}
+
+// getChunkHeader - returns the
+func getChunkHeader(chunkLen int64, signature string) []byte {
+	return []byte(strconv.FormatInt(chunkLen, 16) + ";chunk-signature=" + signature + "\r\n")
+}
+
+// getChunkSignature - returns chunk signature for a given chunk and previous signature.
+func getChunkSignature(chunkData []byte, reqTime time.Time, location, previousSignature, secretAccessKey string) string {
+
+	chunkStringToSign := getChunkStringToSign(reqTime, location, previousSignature, chunkData)
+	signingKey := getSigningKey(secretAccessKey, location, reqTime)
+	return getSignature(signingKey, chunkStringToSign)
+}
+
+// getSeedSignature - returns the seed signature for a given request.
+func (s *StreamingReader) setSeedSignature(req *http.Request) {
+	// Get canonical request
+	canonicalRequest := getCanonicalRequest(*req, ignoredStreamingHeaders)
+
+	// Get string to sign from canonical request.
+	stringToSign := getStringToSignV4(s.reqTime, s.location, canonicalRequest)
+
+	signingKey := getSigningKey(s.secretAccessKey, s.location, s.reqTime)
+
+	// Calculate signature.
+	s.seedSignature = getSignature(signingKey, stringToSign)
+}
+
+// StreamingReader implements chunked upload signature as a reader on
+// top of req.Body's ReaderCloser chunk header;data;... repeat
+type StreamingReader struct {
+	accessKeyID     string
+	secretAccessKey string
+	location        string
```

Comment:
you can keep location as `region` - S3 has blurred the differences but for signature lets call it region for readability.